engine/resources/tar.go

```diff
@@ -0,0 +1,566 @@
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package resources
+
+import (
+	"archive/tar"
+	"bytes"
+	"compress/gzip"
+	"crypto/sha256"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"strings"
+	"sync"
+
+	"github.com/purpleidea/mgmt/engine"
+	"github.com/purpleidea/mgmt/engine/traits"
+	"github.com/purpleidea/mgmt/lang/funcs/vars"
+	"github.com/purpleidea/mgmt/lang/interfaces"
+	"github.com/purpleidea/mgmt/lang/types"
+	"github.com/purpleidea/mgmt/recwatch"
+	"github.com/purpleidea/mgmt/util/errwrap"
+)
+
+const (
+	// KindTar is the kind string used to identify this resource.
+	KindTar = "tar"
+	// ParamTarState is the name of the state field parameter.
+	ParamTarState = "state"
+	// TarStateExists is the string that represents that the tar file should exist
+	TarStateExists = "exists"
+	// TarStateAbsent is the string that represents that the tar file should not exist.
+	TarStateAbsent = "absent"
+)
+
+func init() {
+	engine.RegisterResource(KindTar, func() engine.Res { return &TarRes{} })
+
+	vars.RegisterResourceParams(KindTar, map[string]map[string]func() interfaces.Var{
+		ParamTarState: {
+			TarStateExists: func() interfaces.Var {
+				return &types.StrValue{
+					V: TarStateExists,
+				}
+			},
+			TarStateAbsent: func() interfaces.Var {
+				return &types.StrValue{
+					V: TarStateAbsent,
+				}
+			},
+		},
+	})
+}
+
+// TarRes
+type TarRes struct {
+	traits.Base // add the base methods without re-implementation
+	traits.Edgeable
+	traits.GraphQueryable // allow others to query this res in the res graph
+
+	init *engine.Init
+
+	// Inputs are the input filepaths for the tar resource
+	Inputs []string `lang:"inputs" yaml:"inputs"`
+
+	// State is the desired state of the tar resource. It can either be 'exist' or 'absent'
+	State string `lang:"state" yaml:"state"`
+
+	// Compress is a flag to determine whether to compress the tar archive with
+	// gzip or not. By default it is set to false (which won't compress the archive)
+	Compress bool `lang:"compress" yaml:"compress"`
+
+	// The path to local storage
+	vardir string
+
+	// The tar resource needs to determine if its input files are "dirty":
+	// if so, checkApply will create a new tar resource.
+	// The tar resource uses a hashmap to determine whether files are dirty or not.
+	// Each key is an input file for the tar resource and
+	// each value is a hash of the respective input file's content
+	inputsHashMapFilepath string
+}
+
+// Default returns some sensible defaults for this resource.
+func (obj *TarRes) Default() engine.Res {
+	return &TarRes{}
+}
+
+// Validate if the params passed in are valid data.
+func (obj *TarRes) Validate() error {
+	if len(obj.Inputs) == 0 {
+		return fmt.Errorf("expected one or more inputs")
+	}
+	if obj.Compress && !strings.HasSuffix(strings.ToLower(obj.Name()), ".gz") {
+		return fmt.Errorf("expected a suffix of .gz, .Gz, or .gZ for compressed archive")
+	}
```

Comment:
Could probably remove this. Ideally the user should name the tar file whatever they want.